app/src/main/java/org/p2p/wallet/transaction/ui/JupiterTransactionProgressBottomSheet.kt

```diff
@@ -130,49 +133,90 @@ class JupiterTransactionProgressBottomSheet : BottomSheetDialogFragment() {
         }
     }
 
-    private fun setProgressState(state: TransactionState.Progress) {
-        with(binding) {
-            textViewTitle.text = progressStateFormat.format(getString(R.string.transaction_progress_submitted))
-            progressStateTransaction.setDescriptionText(R.string.transaction_description_progress)
-            buttonDone.setText(R.string.common_done)
+    private fun setErrorState(reason: TransactionStateSwapFailureReason) {
+        when (reason) {
+            is TransactionStateSwapFailureReason.LowSlippage -> setLowSlippageError(reason)
+            is TransactionStateSwapFailureReason.Unknown -> setUnknownTransactionError()
         }
     }
 
-    private fun setErrorState(state: TransactionState.JupiterSwapFailed) {
-        with(binding) {
-            textViewTitle.text = progressStateFormat.format(getString(R.string.transaction_progress_failed))
-            progressStateTransaction.setFailedState()
-            progressStateTransaction.setDescriptionText(R.string.transaction_description_failed)
-            buttonDone.setText(R.string.common_close)
+    private fun setLowSlippageError(error: TransactionStateSwapFailureReason.LowSlippage) = with(binding) {
+        if (error.currentSlippageValue < SLIPPAGE_NEEDED_FOR_MANUAL_CHANGE) {
+            val newSlippage = getNewSlippage(error.currentSlippageValue)
+            progressStateTransaction.setDescriptionText(
+                getString(
+                    R.string.swap_transaction_details_error_low_slippage,
+                    error.currentSlippageValue.toString(),
+                    newSlippage.toString()
+                )
+            )
+
+            dismissResult = JupiterTransactionDismissResult.SlippageChangeNeeded(newSlippage)
+        } else {
+            progressStateTransaction.setDescriptionText(
+                getString(
+                    R.string.swap_transaction_details_error_low_slippage_manual,
+                    error.currentSlippageValue.toString()
+                )
+            )
+            dismissResult = JupiterTransactionDismissResult.ManualSlippageChangeNeeded
         }
+        textViewTitle.text = progressStateFormat.format(getString(R.string.transaction_progress_failed))
+        progressStateTransaction.setFailedState()
+
+        buttonDone.setText(R.string.swap_transaction_details_error_low_slippage_button)
+    }
+
+    private fun setUnknownTransactionError() = with(binding) {
+        dismissResult = JupiterTransactionDismissResult.TrySwapAgain
+
+        textViewTitle.text = progressStateFormat.format(getString(R.string.transaction_progress_failed))
+        progressStateTransaction.setFailedState()
+        progressStateTransaction.setDescriptionText(R.string.swap_transaction_details_error_unknown)
+        buttonDone.setText(R.string.common_try_again)
+    }
+
+    private fun getNewSlippage(currentSlippage: Double): Double = when (currentSlippage) {
+        in (0.0..0.4) -> 0.5
```

Comment:
can we move it so fun or to const or to class?

Reply:
i think i can, but i cant think of any good name for this const
it's very easy mapping from Figma, looks good i think